File: sem1/hw8/task3/hashtable.cpp
#include "mystring.h"
#include "list.h"
#include "hashtable.h"
#include <iostream>

//using namespace std;

const int hashSize = 2000;

struct HashTable{
    List *table[hashSize];
};

HashTable *createHashTable()
{
    HashTable *hash = new HashTable;
    for (int i = 0; i < hashSize; i++)
    {
        hash->table[i] = createList();
    }
    return hash;
}

void deleteHashTable(HashTable *hash)
{
    for (int i = 0; i < hashSize; i++)
    {
        deleteList(hash->table[i]);
    }
    delete hash;
}

int hash(String *string)
{
    long long hash = 2139062143;

    char *sameString = getChar(string);
    int strlen = length(string);

    for (int i = 0; i < strlen; i++)
    {
        hash = (37 * hash + sameString[i]) % hashSize;
    }

    delete [] sameString;
    hash = hash % hashSize;
    return hash;
}

void addToHash(HashTable *table, String *string)
{
    int hashNumber = hash(string);
    add(table->table[hashNumber], string);
}

double amountOfWords(HashTable *hash)
{
    double numberOfWords = 0;
    for (int i = 0; i < hashSize; i++)
    {
        numberOfWords += size(hash->table[i]);
    }
    return numberOfWords;
}

double loadFactor(HashTable *hash)
{
    return (double)(amountOfWords(hash) / hashSize);
}

int maxChainLength(HashTable *hash)
{
    int max = 0;
    for (int i = 0; i < hashSize; i++)
    {
        if (size(hash->table[i]) > max)
            max = size(hash->table[i]);
    }
    return max;
}

void maxChain(HashTable *hash)
{
    int max = 0;
    for (int i = 0; i < hashSize; i++)
    {
        if (size(hash->table[i]) > size(hash->table[max]))
            max = i;
    }
    printList(hash->table[max]);
}

double averageNotEmptyChain(HashTable *hash)
{
    double averageLength = 0;
    int notEmptyChains = 0;
    for (int i = 0; i < hashSize; i++)
    {
        int tempSize = size(hash->table[i]);
        if (tempSize)
        {
            averageLength += size(hash->table[i]);
            notEmptyChains++;
        }
    }
    return (double)(averageLength / notEmptyChains);
}

int emptyCell(HashTable *hash)
{
    int number = 0;
    for (int i = 0; i < hashSize; i++)
    {
        if (size(hash->table[i]) == 0)
            number++;
    }
    return number;
}

int count(HashTable *table, String *string)
{
    int hashNumber = hash(string);
    int result = searchWord(table->table[hashNumber], string);
    return result;
}

File: sеm2/hw4/task2/src/sem2/hw4/task2/HashTable.java
package sem2.hw4.task2;

/** Class for HashTable */
public class HashTable<T extends Comparable>  {

    public int hashSize = 10;
    public UniqueList<T>[] hashTable = new UniqueList[hashSize];
    private Hash hash;

    public HashTable(Hash hash) {
        for (int i = 0; i < hashSize; i++) {
            hashTable[i] = new UniqueList<T>();
        }
        this.hash = hash;
    };

    /**
     * @param value
     * @return hash number for given value
     */
    int hash(T value) {
        String valueString = value.toString();
        return hash.hashNumber(valueString, hashSize);
    }

    /**
     * method that adds an element to hash table.
     *
     * @param value element to be added has this value
     */
    void addToHash(T value) {
        int hashNumber = hash(value);
        hashTable[hashNumber].add(value);
    }

    /**
     * method that removes an element from hash table.
     *
     * @param value element to be deleted has this value
     */
    void removeFromHash(T value) {
        int hashNumber = hash(value);
        hashTable[hashNumber].remove(value);
    }

    /**
     * method that counts amount of different words in hash table.
     *
     * @return amount of different words in hash table
     */
    double amountOfWords() {
        double numberOfWords = 0;
        for (int i = 0; i < hashSize; i++) {
            numberOfWords += hashTable[i].size();
        }
        return numberOfWords;
    }

    /**
     * method that counts load factor of hash table.
     *
     * @return load factor
     */
    double loadFactor() {
        return (double)(amountOfWords() / hashSize);
    }

    /**
     * method that counts max length of chain and returns it.
     *
     * @return max length of chain
     */
    int maxChainLength() {
        int max = 0;
        for (int i = 0; i < hashSize; i++)
        {
            if (hashTable[i].size() > max)
                max = hashTable[i].size();
        }
        return max;
    }

    /**
     * method that counts average length of chain.
     *
     * @return average length
     */
    double averageNotEmptyChain() {
        double averageLength = 0;
        int notEmptyChains = 0;
        for (int i = 0; i < hashSize; i++)
        {
            int tempSize = hashTable[i].size();
            if (tempSize != 0)
            {
                averageLength += hashTable[i].size();
                notEmptyChains++;
            }
        }
        return (double)(averageLength / notEmptyChains);
    }

    /**
     * method that counts the number of conflicts.
     *
     * @return number of conflicts in hash table
     */
    int conflicts() {
        int number = 0;
        for (int i = 0; i < hashSize; i++)
        {
            if (hashTable[i].size() > 1)
                number++;
        }
        return number;
    }

    /**
     * method that counts the quantity of given element.
     *
     * @param value element
     * @return quantity of this element
     */
    int count(T value) {
        int hashNumber = hash(value);
        int result = hashTable[hashNumber].quantity(value);
        return result;
    }

    /** method that prints statistics of hash table */
    void getStatistics() {
        System.out.println("Size: " + hashSize);
        System.out.println("load factor: " + loadFactor());
        System.out.println("number of conflicts: " + conflicts());
        System.out.println("max chain length: " + maxChainLength());
    }
}
